### ISSUE TITLE
Bug report: Comma dangle always throw error

### ISSUE THREAD

**Issue by mondwan** [eslint/eslint#7809]
**Tell us about your environment**

* **ESLint Version:** v3.12.2
* **Node Version:** v6.9.1
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

```js
{
  "globals": {},
  "env": {},
  "rules": {
    "comma-dangle": [
      "error",
      {
        "arrays": "always-multiline",
        "objects": "always-multiline",
        "imports": "always-multiline",
        "exports": "always-multiline",
        "functions": "always-multiline"
      }
    ]
  },
  "parserOptions": {}
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
console.log(
	'a'
);
```

**What did you expect to happen?**

```sh
$> eslint a.js
# No error
```

**What actually happened? Please include the actual, raw output from ESLint.**

```sh
mondwan@mondwan-All-Series:~/Documents/codeTest/babelEslint$ eslint --print-config .
{
  "globals": {},
  "env": {},
  "rules": {
    "comma-dangle": [
      "error",
      {
        "arrays": "always-multiline",
        "objects": "always-multiline",
        "imports": "always-multiline",
        "exports": "always-multiline",
        "functions": "always-multiline"
      }
    ]
  },
  "parserOptions": {}
}
mondwan@mondwan-All-Series:~/Documents/codeTest/babelEslint$ cat a.js 
console.log(
	'a'
);
mondwan@mondwan-All-Series:~/Documents/codeTest/babelEslint$ eslint a.js 

/home/mondwan/Documents/codeTest/babelEslint/a.js
  2:5  error  Missing trailing comma  comma-dangle

✖ 1 problem (1 error, 0 warnings)


```

**Comment by not-an-aardvark** [comment]
Hi,

This looks like it's working as intended. You have the `functions` option set to `always-multiline`, so you should add a trailing comma like this:

```js
console.log(
        'a',
);
```

Alternatively, if you want to disallow trailing commas for functions (trailing function commas aren't supported yet in some engines), you can configure the rule like this:

```json
{
  "rules": {
    "comma-dangle": [
      "error",
      {
        "arrays": "always-multiline",
        "objects": "always-multiline",
        "imports": "always-multiline",
        "exports": "always-multiline",
        "functions": "never"
      }
    ]
  }
}
```

**Comment by mondwan** [comment]
Really? Does ES6 allows us to type arguments in this manner?

As ES5, `console.log('a',)` is a syntax error.

Although the rule's name tells what it suppose to check, I don't think it
is good to show a warning in this scenario which lead to a syntax error.



2016年12月23日 12:05，"Teddy Katz" <notifications@github.com>寫道：

**Comment by not-an-aardvark** [comment]
Trailing function commas are allowed in ES8 syntax.

By default, the rule will not report an error for missing function commas. However, since your configuration contains `"functions": "always-multiline"`, it reports an error. To avoid this, you should use `"functions": "never"` (or omit the `"functions"` option), and the rule will no longer report an error.

**Comment by mondwan** [comment]
Oh, thanks. I have tried `never` in `functions` and it works as expected.

Just one more question. Does `parserOption.ecmaVersion` means anything here? As I have tried explicitly specified this `ecmaVersion: 5` along with `function: always-multiline`, it still shows an error.

Should `ecmaVersion` controls how other rules behaves?

Configuration here
```json
{
  "globals": {},
  "env": {},
  "rules": {
    "comma-dangle": [
      "error",
      {
        "arrays": "always-multiline",
        "objects": "always-multiline",
        "imports": "always-multiline",
        "exports": "always-multiline",
        "functions": "always-multiline"
      }
    ]
  },
  "parserOptions": {
    "ecmaVersion": 5
  }
}
```

**Comment by not-an-aardvark** [comment]
No, `ecmaVersion` generally does not affect how rules behave. To prevent missing commas from being reported, you should set the `functions` option explicitly (or omit the option, since the default behavior is to ignore trailing function commas).

**Comment by mondwan** [comment]
Will this feature be implemented in the future?

Since syntax may be keep varied significantly as this issue revealed in the
coming future, it may break existing linting results even users are still
using ES5 at all.

Just an open question and thanks for answering. Merry Christmas:)

Close #7809

2016年12月26日 13:25，"Teddy Katz" <notifications@github.com>寫道：

No, ecmaVersion generally does not affect how rules behave. To prevent
missing commas from being reported, you should set the functions option
explicitly (or omit the option, since the default behavior is to ignore
trailing function commas).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/eslint/eslint/issues/7809#issuecomment-269164066>, or mute
the thread

**Comment by ljharb** [comment]
It definitely seems like trailing function commas should never be warned prior to an `ecmaVersion` of 2017, since there's no fix that's correct there.

**Comment by mysticatea** [comment]
People can use ESLint with a custom parser, and `parserOptions` are defined by the each custom parser. For example, `babel-eslint` does not have `parserOptions.ecmaVersion` option.

So, core rules do not depend on `parserOptions` basically.

**Comment by ljharb** [comment]
Interesting, that does make sense.

Perhaps the rule should use the current parser, and determine if the dangling comma would be a syntax error or not, before requiring it? That seems like it would work for every parser.

**Comment by not-an-aardvark** [comment]
I don't think this is a problem, because the default behavior is to ignore function commas (even when `comma-dangle: [error, always]` is used. The rule only checks for trailing commas in functions if the user explicitly enables `functions: always` in their config. If someone is explicitly opting into requiring commas for functions, then it seems unnecessary to use things like parser heuristics to determine whether they actually mean it.

**Comment by ljharb** [comment]
Fair point :-)

**Comment by ljharb** [comment]
Maybe it would still be useful to do one require-time check for the syntax, and error out if the syntax is not supported and it is enabled - that way users will know if it's an option.

One way to do that could be making the schema conditional on whether the parser allows the syntax.

**Comment by mondwan** [comment]
In my case, I think this is a bug until @not-an-aardvark answer. It is not very straight forward for me to think this is a kind of `ECMA` versioning problem as everything work fine before upgrading `eslint` from some old version.

That's why I have such question about `ecmaVersion`. If errors include some message about `ECMA` version, I maybe able to debug it myself at the first place.

**Comment by platinumazure** [comment]
I could get behind a messaging change that mentions when dangling commas in functions are actually valid. @not-an-aardvark Thoughts?

**Comment by not-an-aardvark** [comment]
I would be fine with it, but I'm a bit confused about how it would be accomplished. I suppose having the rule call the user's parser would work.

**Comment by platinumazure** [comment]
Oh, I wouldn't want to add a parser call- I just meant improving the lint
message to say that dangling commas in function arguments are only allowed
in ES2017. Sorry for the confusion.

**Comment by ljharb** [comment]
I'm suggesting exactly that - have the rule call the user's parser with `function test(a,) {}` and see if it successfully parses - if so, great, if not, error out when `functions` is used.

**Comment by ljharb** [comment]
"only allowed in ES2017" wouldn't be quite accurate, since as you say, people could be using a custom parser - that message would only be valid for the default parser.

**Comment by platinumazure** [comment]
@ljharb I was trying to reference a standard, not an `ecmaVersion`. But it might be a bit too subtle a distinction, so I'm open to suggestions.

**Comment by ljharb** [comment]
What would be the problem with a single parser call at rule load time, cached at module level?

**Comment by platinumazure** [comment]
I don't have a position (yet) on attempting to parse. I was only trying to clarify what I was asking @not-an-aardvark the whole time :smile:

**Comment by nzakas** [comment]
As a guideline, we don't want rules re-parsing code or using the `ecmaVersion` to change their behavior.

I don't think there's a change in the rule that makes sense here. I think we could update the docs for the rule to make it more obvious that the `functions` option only applies to ES2017 and higher.

**Comment by losingkeys** [comment]
I think I came across this issue when saving [this file](https://github.com/exercism/xjavascript/blob/401e8cb76c71b9d7b247fca0bdcacaad59b85c01/exercises/hamming/hamming.spec.js#L36). It looks like this is pretty well understood already, but if you need another example, that link provides a good one (afaict).

**Comment by mondwan** [comment]
Umm, I do not see connections between the file you link and issue here. Would you mind to elaborate on how does it links to issue here?

**Comment by losingkeys** [comment]
Ohh, maybe I misread the issue then. When I use `eslint --fix` on the file, eslint adds a trailing comma to [this line](https://github.com/exercism/xjavascript/blob/401e8cb76c71b9d7b247fca0bdcacaad59b85c01/exercises/hamming/hamming.spec.js#L36), which gives me an error when running jasmine in node 7.6.0. Here's my `.eslintrc.js`:
```js
module.exports = {
  extends: 'airbnb',
  parserOptions: {
    ecmaVersion: 2017,
    ecmaFeatures: {
      commaDangle: true,
      impliedStrict: true,
      jsx: true,
      modules: true,
      module: true,
    },
    env:{
      es6: true
    },
  },
  plugins: [
    'react',
  ],
};
```

**Comment by ljharb** [comment]
@losingkeys node 7.6 doesn't support trailing function commas, so if you're not using babel, you can't use the `functions` setting.
